File: chapters/physics_solvers/verlet/code/julia/verlet.jl
function verlet(pos::Float64, acc::Float64, dt::Float64)
    prev_pos = pos
    time = 0.0

    while (pos > 0)
        time += dt
        temp_pos = pos
        pos = pos * 2 - prev_pos + acc * dt * dt
        prev_pos = temp_pos
    end

    println(time)
end

function stormer_verlet(pos::Float64, acc::Float64, dt::Float64)
    prev_pos = pos
    time = 0.0
    vel = 0.0

    while (pos > 0.0)
        time += dt
        temp_pos = pos
        pos = pos * 2 - prev_pos + acc * dt * dt
        prev_pos = temp_pos

        # Because acceleration is constant, velocity is straightforward
        vel += acc*dt
    end

    println(time)
end

function velocity_verlet(pos::Float64, acc::Float64, dt::Float64)
    prev_pos = pos
    time = 0.0
    vel = 0.0

    while (pos > 0.0)
        time += dt
        pos += vel * dt + 0.5 * acc * dt * dt;
        vel += acc * dt;
    end

    println(time)
end

function main()
    verlet(5.0, -10.0, 0.01);
    stormer_verlet(5.0, -10.0, 0.01);
    velocity_verlet(5.0, -10.0, 0.01);
end

main()

File: chapters/physics_solvers/verlet/code/rust/verlet.rs
fn verlet(mut pos: f64, acc: f64, dt: f64) {
    let mut prev_pos = pos;
    let mut time = 0.0;

    while pos > 0.0 {
        time += dt;
        let temp_pos = pos;
        pos = pos * 2.0 - prev_pos + acc * dt * dt;
        prev_pos = temp_pos;
    }

    println!("{}", time);
}

fn stormer_verlet(mut pos: f64, acc: f64, dt: f64) {
    let mut prev_pos = pos;
    let mut time = 0.0;

    while pos > 0.0 {
        time += dt;
        let temp_pos = pos;
        pos = pos * 2.0 - prev_pos + acc * dt * dt;
        prev_pos = temp_pos;
    }

    println!("{}", time);
}

fn velocity_verlet(mut pos: f64, acc: f64, dt: f64) {
    let mut time = 0.0;
    let mut vel = 0.0;

    while pos > 0.0 {
        time += dt;
        pos += vel * dt + 0.5 * acc * dt * dt;
        vel += acc * dt;
    }

    println!("{}", time);
}

fn main() {
    verlet(5.0, -10.0, 0.01);
    stormer_verlet(5.0, -10.0, 0.01);
    velocity_verlet(5.0, -10.0, 0.01);
}
